Added FastClonerEnumMap and FastClonerLinkedHashSet

tweimer opened the pull request:
Hi Konstantinos,
I am using your Cloning library, and I found a bug with cloning of EnumMaps.
If any key of an EnumMap is being mapped to `null`, the cloned EnumMap will *not* map it to `null`, but to a random instance of `Object`.

I tracked down that problem and the cause of the problem is that `EnumMap` is using a marker object `java.util.EnumMap.NULL` to distinguish enums mapped to `null` from objects not contained in the map at all. `java.util.EnumMap.NULL` is used in `maskNull` and `unmaskNull` by reference comparision, so after `NULL` is cloned, the reference comparision goes wrong.

You are able to reproduce that problem with the JUnit test included in my Pull Request.

In order to solve the problem, I have introduced `FastClonerEnumMap`. It is basically taken from `FastClonerHashMap`, with some modification for `EnumMap`. It is registered in `com.rits.cloning.Cloner.registerFastCloners()`.

After that, the JUnit tests I added are working fine. Feel free to add additional ones.

I also added FastClonerLinkedhashSet, which will fix #46 (JUnit tests are included too).

I also updated pom.xml (fixes #66)

I hope you will merge this PR. If you have any objections, please let me know!

kostaskougios commented on src/test/java/com/rits/tests/cloning/TestCloner.java:
```diff
@@ -796,6 +796,7 @@ public void testEnumMapWithNullValue() {
 		originalMap.put(TestEnum.A, null);
 
 		EnumMap<TestEnum, String> clonedMap = cloner.deepClone(originalMap);
+		assertTrue("Cloned LinkedHashSet not instanceof EnumMap", clonedMap instanceof EnumMap);
```
Why do we need this assertion here? If clonedMap wasn't an EnumMap, wouldn't we get a class cast exception on previous line (798)?

tweimer replied:
Well, reading that description I just copy-pasted it from the lines below. :) I agree this is redundant, I'll remove that.

tweimer replied:
Fixed.

kostaskougios commented on src/test/java/com/rits/tests/cloning/TestCloner.java:
```diff
@@ -790,4 +790,60 @@ public void testConcurrentLinkedQueue() {
+	public void testEmptyEnumMap() {
+		EnumMap<TestEnum, String> originalMap = new EnumMap<>(TestEnum.class);
+		EnumMap<TestEnum, String> clonedMap = cloner.deepClone(originalMap);
+		assertTrue("Cloned LinkedHashSet not instanceof EnumMap", clonedMap instanceof EnumMap);
```
^^^ and here, classcast on 808

tweimer replied:
Agreed, see above.

tweimer replied:
Fixed.

kostaskougios commented on src/test/java/com/rits/tests/cloning/TestCloner.java:
```diff
@@ -790,4 +790,60 @@ public void testConcurrentLinkedQueue() {
+
+	public void testEnumMapWithNonNullValue() {
+		EnumMap<TestEnum, String> originalMap = new EnumMap<>(TestEnum.class);
+		originalMap.put(TestEnum.A, "Hello");
```
One case that is not tested is when the values are actually a non-immutable class. Will those be cloned correctly? (Looking at the EnumMapFastCloner code they are cloned, but maybe a test to prove so would be useful in the future)

tweimer replied:
Agreed, I will add it.

tweimer replied:
Added.

kostaskougios commented on src/test/java/com/rits/tests/cloning/TestCloner.java:
```diff
@@ -790,4 +790,60 @@ public void testConcurrentLinkedQueue() {
+	
+	public void testLinkedHashSetEquals() {
+		Set<Object> set = new LinkedHashSet<>();
+		set.add("Test 1");
```
A test that uses mutable values for the LinkedHashSet would be useful to prove that mutable values are indeed cloned.

tweimer replied:
Agreed, I will add it.

tweimer replied:
Added.

tweimer commented:
> Hi Tobias, thanks for the changes. My review comments follows. Please let me know if you want to change anything or you're happy for your code to be merged.

Thamks. I fixed/added all your remarks and added even more JUnit. :) Ready to be merged!

tweimer commented:
Thank you! Could you make a new release soon? I really need that fix and I wouldn't want to use my own workaround.

kostaskougios commented:
Hi, just deployed 1.10.2 with the fixes
